lib/chef/provider/remote_directory.rb

```diff
@@ -161,7 +161,7 @@ def purge_file(file)
       def files_to_transfer
         cookbook = run_context.cookbook_collection[resource_cookbook]
         files = cookbook.relative_filenames_in_preferred_directory(node, :files, source)
-        files.sort!.reverse!
+        files.sort_by! { |x| x.split(::File::SEPARATOR).count }
```

Comment:
can we just `x.count(::File::SEPARATOR)` instead of splitting it?  that should provide a large speed boost and this is fairly speed critical code if you've got tons of files in the directory.

also, should check on windows that this still works correctly, but i think this is right (IIRC, `::File::SEPARATOR` on Windows is still '/' and we should be passing those back to ourselves -- but if not, we'd need to use `ALT_SEPARATOR` instead...)